Update to net8.0 SDK

wtgodbe opened the pull request.

wtgodbe committed:
Update to net8.0 SDK

wtgodbe commented:
Blocked on getting an SDK from https://github.com/dotnet/installer/pull/15151

mitchdenny reviewed (approved):
Change looks good (once you've got the installer available).

dougbu commented on eng/tools/RepoTasks/RepoTasks.csproj:
```diff
@@ -1,6 +1,6 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>net7.0</TargetFrameworks>
+    <TargetFrameworks>$(DefaultNetCoreTargetFramework)</TargetFrameworks>
```
Changes here cannot work until we have an 8.0.0 SDK containing `net8.0` bits. The basic idea is `RepoTasks` and `GenerateFiles` aren't part of the product. They're instead used in `msbuild` (or `dotnet msbuild`) and to generate the extra Directory.Build.props, Directory.Build.targets, and dotnet-tools.json files used in almost every other project in the repo.

Why did this file and RepoTasks.tasks need to change in the first place❔ Note `GenerateFiles` still targets `net5.0` and (other than the deprecated TFM warning we're suppressing somewhere) that's fine.

See for example
``` text
System.IO.FileNotFoundException: Could not load file or assembly 'System.Runtime, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.
```

wtgodbe replied:
Yep, that's why this is blocked on https://github.com/dotnet/installer/pull/15151. It needs to change because SourceBuild breaks if it doesn't target the latest TFM (though I don't know why the same isn't true for `GenerateFiles`) - https://github.com/dotnet/installer/pull/15151#issuecomment-1359316105

ViktorHofer replied:
> though I don't know why the same isn't true for GenerateFiles

@dougbu @wtgodbe, GenerateFiles.csproj doesn't need to use the latest TFM because it doesn't build during source-build: https://github.com/dotnet/aspnetcore/blob/cefc6ccf4b23ac9f07418cd10e374cf533276eee/eng/tools/GenerateFiles/GenerateFiles.csproj#L5

wtgodbe committed:
Update global.json

wtgodbe commented:
Here's a new one, will look tomorrow:

> /Users/runner/work/1/s/.dotnet/sdk/8.0.100-alpha.1.23054.7/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(90,5): error NETSDK1195: Unable to optimize assemblies for size: a valid runtime package was not found. Either set the PublishTrimmed property to false, or use a supported target framework when publishing. [/Users/runner/work/1/s/src/Servers/Connections.Abstractions/src/Microsoft.AspNetCore.Connections.Abstractions.csproj::TargetFramework=netstandard2.1]

wtgodbe commented:
@JamesNK can we just disable trimming for `netstandard2.1` targets?

JamesNK commented:
What do you mean by disable trimming? Our projects are annotated to help trimming succeed, but that's it.

Trimming happens when publishing an app. Who is publishing?

mitchdenny commented:
> Who is publishing?

I think our Helix tests publish. Not sure if they would be impacted by this.

wtgodbe commented:
> Trimming happens when publishing an app. Who is publishing?

This error happens very early on, during framework reference resolution. Presumably the SDK is doing some scanning that sees we have `PublishTrimmed=true` for projects with a `netstandard2.1` configuration, and errors out. I'm proposing setting that to false whenever `TargetFramework=netstandard2.1`

wtgodbe commented:
@sbomer looks like this was caused by https://github.com/dotnet/sdk/pull/29441 - is it intentional that we get these errors for netstandard2.1 projects that enable trimming?

https://github.com/dotnet/aspnetcore/pull/45879#issuecomment-1371754009

wtgodbe commented:
> @wtgodbe Is it definitely the case that PublishTrimmed is set, or could these projects have IsTrimmable or EnableTrimAnalyzer set instead? netstandard2.1 didn't have annotations for trimming, but the 7.0 SDK didn't block enabling the trim analyzer, and it was possible to get warnings by for example compiling a copy of RequiresUnreferencedCodeAttribute into user code.

They set `IsTrimmable`. Maybe these projects could conditionally set `PublishTrimmed=false` when `TargetFramework=netstandard2.1`

https://github.com/dotnet/aspnetcore/blob/5b43ed1b8355c9fdebb44b0035d215a563435562/src/Servers/Connections.Abstractions/src/Microsoft.AspNetCore.Connections.Abstractions.csproj#L10

wtgodbe commented:
> Setting IsTrimmable to false would prevent the library from being trimmed when consumed in apps, so it may not be what you want. I think our team needs to discuss the intended UX around this. For now, I can suggest a workaround:

I'm suggesting keeping `IsTrimmable=true` everywhere, but also adding `PublishTrimmed=false` when TFM=netstandard2.1. Would that not work? If not I'll try your suggested workaround

wtgodbe committed:
Update Directory.Build.targets.in

wtgodbe committed:
Update RedisHubLifetimeManager.cs

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/arcade build 20230105.3

Microsoft.DotNet.Arcade.Sdk , Microsoft.DotNet.Build.Tasks.Installers , Microsoft.DotNet.Build.Tasks.Templating , Microsoft.DotNet.Helix.Sdk
 From Version 8.0.0-beta.22630.1 -> To Version 8.0.0-beta.23055.3

brunolins16 commented on global.json:
```diff
@@ -1,9 +1,9 @@
+    "version": "8.0.100-alpha.1.23054.7"
   },
   "tools": {
-    "dotnet": "8.0.100-alpha.1.22531.1",
```
Because of the new SDK, I believe we will need to update the CI pipeline to use the `binlogtosln` that supports `v15` (https://github.com/dotnet/arcade/issues/11995), if it is not updated yet.

wtgodbe replied:
Just merged that PR into this one

wtgodbe committed:
Update Arcade

wtgodbe committed:
Fix conflict

wtgodbe committed:
NoWarn

wtgodbe commented:
Remaining errors are all:

> ##[error]src/Tools/LinkabilityChecker/LinkabilityChecker.csproj(47,5): error MSB4036: (NETCORE_ENGINEERING_TELEMETRY=Build) The "ILLink" task was not found. Check the following: 1.) The name of the task in the project file is the same as the name of the task class. 2.) The task class is "public" and implements the Microsoft.Build.Framework.ITask interface. 3.) The task is correctly declared with <UsingTask> in the project file, or in the *.tasks files located in the "/Users/runner/work/1/s/.dotnet/sdk/8.0.100-alpha.1.23055.1" directory.

@sbomer are you aware of any other recent linker changes that might have caused this to start happening?

dougbu commented:
Removing ILLink sounds like a regression, especially if the SDK doesn't treat references to the SDK in the same way it does missing targeting packs. What's the mid-term story here❔

JamesNK commented:
@dougbu It sounds like ILLink isn't in the SDK by default, but it is automatically downloaded when `<PublishTrimmed>true</PublishTrimmed>` is set on a project. What we're doing here, calling the ILLink task directly, is something people would rarely do and is an acceptable tradeoff for a smaller SDK size.

All we should need to do is add `<PublishTrimmed>true</PublishTrimmed>` to `WasmLinkerTest.csproj` and `LinkabilityChecker.csproj`.

@sbomer Correct me if I'm wrong here.

wtgodbe committed:
Update WasmLinkerTest.csproj

wtgodbe committed:
Update LinkabilityChecker.csproj

ViktorHofer reviewed (approved):
LGTM. If you want to avoid this for .NET 9 and onwards, consider setting `DisableImplicitFrameworkReferences` to true in RepoTasks.csproj as presumably that project doesn't need the framework references and the underlying targeting pack.

wtgodbe commented:
> f you want to avoid this for .NET 9 and onwards, consider setting DisableImplicitFrameworkReferences to true in RepoTasks.csproj as presumably that project doesn't need the framework references and the underlying targeting pack.

Thanks, testing that out pre-SDK update here: https://github.com/dotnet/aspnetcore/pull/45971

wtgodbe commented:
Don't think that'll work since our RepoTasks do use stuff from runtime. We can probably just apply the FrameworkReference workaround we give to non-Tools projects to RepoTasks to fix this going forward.

ViktorHofer commented:
I don't think we need to fix this right now and would instead love to see this go in. Should we file an issue to track the follow-up work?

wtgodbe commented:
https://github.com/dotnet/aspnetcore/issues/45973

dougbu commented:
@wtgodbe this change broke official builds. See <https://dev.azure.com/dnceng/internal/_build/results?buildId=2083576&view=results> for example. Please immediately either revert this change or update the `binlogtosln` tool version we use.

wtgodbe commented:
I see, this did bring in the new SourceIndexPackageVersion from https://github.com/dotnet/arcade/issues/11995, but we overwrite that in ci.yml with an old version from 2021. We should probably just start using the default from Arcade. I'll open a PR now

wtgodbe commented:
Never mind, can't remove that because it's used directly before we get to eng/common. Opened https://github.com/dotnet/aspnetcore/pull/45982 to just update it
